hail/python/hailtop/aiocloud/aiogoogle/client/storage_client.py

```diff
@@ -639,22 +640,76 @@ def schemes() -> Set[str]:
     def storage_location(self, uri: str) -> str:
         return self.get_bucket_and_name(uri)[0]
 
-    async def is_hot_storage(self, location: str) -> bool:
+    async def is_hot_storage(self, uri: str) -> bool:
         """
         See `the GCS API docs https://cloud.google.com/storage/docs/storage-classes`_ for a list of possible storage
         classes.
 
         Raises
         ------
         :class:`aiohttp.ClientResponseError`
-            If the specified bucket does not exist, or if the account being used to access GCS does not have permission
-            to read the bucket's default storage policy.
+            If the specified object does not exist, or if the account being used to access GCS does not have permission
+            to read the bucket's default storage policy and it is not a public access bucket.
         """
-        return (await self._storage_client.bucket_info(location))["storageClass"].lower() in (
-            "standard",
-            "regional",
-            "multi_regional",
-        )
+        is_hot_storage = False
+        hot_storage_classes = {"standard", "regional", "multi_regional"}
+        errors: List[aiohttp.ClientResponseError] = []
+        location = self.storage_location(uri)
+
+        async def is_bucket_hot() -> bool:
+            try:
+                return (await self._storage_client.bucket_info(location))["storageClass"].lower() in hot_storage_classes
+            except aiohttp.ClientResponseError as e:
+                errors.append(e)
+                return False
+
+        async def is_object_hot() -> bool:
+            try:
+                return (await (await self.statfile(uri))["storageClass"]).lower() in hot_storage_classes
+            except aiohttp.ClientResponseError as e:
+                errors.append(e)
+                return False
+            except FileNotFoundError:
+                return False
+
+        async def is_dir_first_object_hot() -> bool:
+            try:
+                files = await self.listfiles(uri, recursive=True)
+                next_file = await files.__anext__()
+                while await next_file.is_dir():
+                    next_file = await files.__anext__()
+                return await self.is_hot_storage(await next_file.url())
```

Comment:
I think this can be written more idiomatically using an `async for`:

```python
async for entry in await self.listfiles(uri, recursive=True):
    if await entry.is_file():
        return await self.is_hot_storage(await file.url())
```

Reply:
updated!

Reply:
updated!